### PR TITLE
Added deleted-collections endpoint for harvesting client

### PR DESCRIPTION
**Pull request by ygliuvt**
To support metadata harvesting, a harvesting client can search CMR for collections that are deleted within a time period. The only search parameter supported is `revision_date` and its format is the same as the `revision_date` parameter in regular collection search. The only supported result format is xml references. The response is the references to the highest non-tombstone collection revisions of the collections that are deleted within the given revision date period.

### CODE REVIEW

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj [hunk 1]
```diff
@@ -666,3 +666,33 @@
...
+    :revision_date
+    :result-format})
+
+(defn- unrecognized-deleted-colls-params-validation
+  "Validates that no invalid parameters were supplied to deleted collections search"
+  [params]
+  (map #(format "Parameter [%s] was not recognized." (csk/->snake_case_string %))
+       (set/difference (set (keys params)) valid-deleted-collections-search-params)))
+
+(defn- deleted-colls-result-format-validation
```
Comment (indiejames) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: Was it in the requirement to return XML? Just curious because it seems like our json response is so much easier to work with.

File: search-app/src/cmr/search/api/routes.clj [hunk 2]
```diff
@@ -369,11 +384,11 @@
 
         ;; add routes for managing jobs
         (common-routes/job-api-routes
-          (routes
-            (POST "/refresh-collection-metadata-cache" {:keys [headers params request-context]}
-              (acl/verify-ingest-management-permission request-context :update)
-              (metadata-cache/refresh-cache request-context)
-              {:status 200})))
+         (routes
```
Comment (indiejames) [inline search-app/src/cmr/search/api/routes.clj]: Reformatting just to get extra lines in your commit? That's cheating, Leo. :)

File: search-app/src/cmr/search/services/query_service.clj [hunk 3]
```diff
@@ -301,6 +304,102 @@
      (ph/provider-holdings->string
        (:result-format params) provider-holdings {:echo-compatible? (= "true" echo-compatible)})]))
 
+(defn- get-collections-with-deleted-revisions
+  "Executes elasticsearch searches to find collections that have deleted revisions.
+   Returns a list of concept-ids of the collections."
+  [context params]
+  ;;1/ Find all collection revisions that are deleted satisfy the original search params
```
Comment (indiejames) [inline search-app/src/cmr/search/services/query_service.clj]: Is the 1/ a typo?

File: search-app/src/cmr/search/services/query_service.clj [hunk 4]
```diff
@@ -301,6 +304,102 @@
...
+          (assoc :hits (count highest-revisions))))))
+
+(defn get-deleted-collections
+  "Executes elasticsearch searches to find collections that are deleted.
+   This only finds collections that are truely deleted and not searchable.
+   Collections that are deleted, then later ingested again are not included in the result.
+   Returns references to the highest collection revision prior to the collection tombstone."
+  [context params]
+  (pv/validate-deleted-collections-params params)
+  ;; 1/ Find all collection revisions that are deleted after satify the revision-date query -> c1
```
Comment (indiejames) [inline search-app/src/cmr/search/services/query_service.clj]: satisfy typo

File: search-app/src/cmr/search/services/query_service.clj [hunk 5]
```diff
@@ -301,6 +304,102 @@
...
+                                              (assoc :page-size :unlimited)))
+        coll-concept-ids (map #(first (:concept-id (:fields %)))
+                              (get-in results [:hits :hits]))]
+    (distinct coll-concept-ids)))
+
+(defn- get-visible-collections
+  "Returns the concept ids of collections that are visible from the given collection concept ids"
+  [context coll-concept-ids]
+  (when (seq coll-concept-ids)
+    (let [condition (qm/string-conditions :concept-id coll-concept-ids true)
```
Comment (indiejames) [inline search-app/src/cmr/search/services/query_service.clj]: I think we use boolean OR conditions to generate the condition for qm/string-condtions (we probably shouldn't). I believe there is a limit of 1024 clauses, which means we can only specify 1024 concept-ids here. It seems like that might be a problem.

File: system-int-test/src/cmr/system_int_test/utils/search_util.clj [hunk 6]
```diff
@@ -1,40 +1,40 @@
...
-    [cmr.system-int-test.utils.dev-system-util :as dev-util]
-    [cmr.system-int-test.utils.fast-xml :as fx]
-    [cmr.system-int-test.utils.url-helper :as url]
-    [cmr.transmit.config :as transmit-config]
-    [cmr.umm.dif.dif-collection]
-    [cmr.umm.iso-mends.iso-mends-collection]
-    [cmr.umm.iso-smap.iso-smap-collection]
-    [ring.util.codec :as codec]))
+  (:require
+   [camel-snake-kebab.core :as csk]
```
Comment (indiejames) [inline system-int-test/src/cmr/system_int_test/utils/search_util.clj]: More cheats!

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj (hunk 1 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: CMR only use xml format for returning collection references, we don't use json for that.

File: search-app/src/cmr/search/api/routes.clj (hunk 2 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/api/routes.clj]: Not intentional. :) And I think github should be smart enough to ignore whitespaces.

File: search-app/src/cmr/search/services/query_service.clj (hunk 3 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_service.clj]: It was from the `get-deleted-collections` function where it has steps 1/, 2/, 3/. I intentionally left it here to match that, but I think I will take it off now that it confuses you.

File: search-app/src/cmr/search/services/query_service.clj [hunk 7]
```diff
@@ -301,6 +304,102 @@
...
+                             #((set highest-coll-revisions)
+                               [(:concept-id %) (:revision-id %)])
+                             (:items results))]
+      (-> results
+          (assoc :items highest-revisions)
+          (assoc :hits (count highest-revisions))))))
+
+(defn get-deleted-collections
+  "Executes elasticsearch searches to find collections that are deleted.
+   This only finds collections that are truely deleted and not searchable.
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/query_service.clj]: truly

File: search-app/docs/api.md [hunk 8]
```diff
@@ -2473,6 +2474,48 @@ __Example Response__
 }
 ```
 
+### <a name="deleted-collections"></a> Find collections that have been deleted
+
+To support metadata harvesting, a harvesting client can search CMR for collections that are deleted within a time period. The only search parameter supported is `revision_date` and its format is the same as the `revision_date` parameter in regular collection search. The only supported result format is xml references. The response is the references to the highest non-tombstone collection revisions of the collections that are deleted within the given revision date period. e.g.
```
Comment (chris-durbin) [inline search-app/docs/api.md]: How much work would it be to support all of the same formats as the collections route?

File: search-app/src/cmr/search/services/query_service.clj [hunk 9]
```diff
@@ -301,6 +304,102 @@
...
+                           :condition condition
+                           :all-revisions? true
+                           :page-size :unlimited
+                           :result-format result-format})
+          results (qe/execute-query context query)
+          highest-coll-revisions (u/map-values
+                                  #(apply max (map :revision-id %))
+                                  (group-by :concept-id (:items results)))
+          highest-revisions (filter
+                             #((set highest-coll-revisions)
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/query_service.clj]: Probably easier to read if changing to (fn[] since it covers multiple lines. It took me a few reads to really get what this code was doing using set as a function to look to see if the given concept revision existed in the set.

File: search-app/src/cmr/search/services/query_service.clj [hunk 10]
```diff
@@ -301,6 +304,102 @@
...
+        visible-concept-ids (get-visible-collections context coll-concept-ids)
+        ;; Find the concept ids that are still deleted
+        deleted-concept-ids (seq (set/difference
+                                   (set coll-concept-ids)
+                                   (set visible-concept-ids)))
+        results (get-highest-visible-revisions context deleted-concept-ids result-format)
+        ;; when results is nil, hits is 0
+        results (or results {:hits 0 :items []})
+        total-took (- (System/currentTimeMillis) start-time)
+        ;; construt the response results string
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/query_service.clj]: construct typo

File: search-app/src/cmr/search/services/query_service.clj (hunk 5 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_service.clj]: We don't have any special code for handling concept-id searches, i.e. this would be the same kind of condition for searching by concept-ids. I tried a test that searches for collections with 2000 concept-ids and it didn't cause any errors. I think we can address this later if we see real problems.

File: search-app/src/cmr/search/services/query_service.clj (hunk 4 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_service.clj]: Done.

File: search-app/src/cmr/search/services/query_service.clj [hunk 11]
```diff
@@ -301,6 +304,102 @@
...
+        results-str (common-search/search-results->response
+                     context
+                     ;; pass in a fake query to get the desired response format
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/query_service.clj]: I don't understand this part about the fake query and the results-str. What does this have to do with the response format?

File: search-app/src/cmr/search/services/query_service.clj [hunk 12]
```diff
@@ -301,6 +304,102 @@
...
+                     (qm/query {:concept-type :collection
+                                :result-format result-format})
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/query_service.clj]: Were you planning on eventually supporting other results formats?

File: search-app/src/cmr/search/services/query_service.clj (hunk 5 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_service.clj]: After reading the code a bit more, the OR string conditions are not turned into boolean or conditions, they are turned to term condition with multiple values. So there is no limit on that as the test has proven.

File: system-int-test/src/cmr/system_int_test/utils/search_util.clj (hunk 6 above)
Comment (ygliuvt) [inline system-int-test/src/cmr/system_int_test/utils/search_util.clj]: Again not intentional. :)

File: search-app/src/cmr/search/services/query_service.clj (hunk 9 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_service.clj]: Done.

File: search-app/src/cmr/search/services/query_service.clj (hunk 7 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_service.clj]: Done.

File: search-app/src/cmr/search/services/query_service.clj (hunk 10 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_service.clj]: Done.

File: search-app/src/cmr/search/services/query_service.clj (hunk 11 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_service.clj]: The `search-results->response` function takes a query argument and use it to determine the response body that it will generate. This fake query is to set up the fields that will be used to generate the correct response.

File: search-app/src/cmr/search/services/query_service.clj (hunk 12 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_service.clj]: No.

File: search-app/docs/api.md (hunk 8 above)
Comment (ygliuvt) [inline search-app/docs/api.md]: As I have commented on the previous comment, we have not supported any formats other than xml for all revisions search. Most of the consideration is probably performance related. It could become really expensive to support more heavyweight formats with all revisions search. For now, we don't have a use case to have to do this. We can talk about it if you really want to look into it.
